src/back/spv/writer.rs

```diff
@@ -1318,6 +1400,494 @@ impl Writer {
+        // in the module's declarations section at the front. In this case,
+        // SPIR-V lets us omit the empty 'else' block, and branch directly to
+        // the merge block. The phi instructions in the merge block can cite the
+        // header block as its CFG predecessor.
+        block.body.push(Instruction::selection_merge(
+            merge_block,
+            spirv::SelectionControl::NONE,
+        ));
+        function.consume(
+            std::mem::replace(block, Block::new(in_bounds_block)),
+            Instruction::branch_conditional(condition, in_bounds_block, merge_block),
+        );
+
+        // The in-bounds path. Perform the access and the load.
+        let value_id = emit_load(&mut self.id_gen, block);
+
+        // Finish the in-bounds block and start the merge block. This
+        // is the block we'll leave current on return.
+        function.consume(
+            std::mem::replace(block, Block::new(merge_block)),
+            Instruction::branch(merge_block),
+        );
+
+        // For the out-of-bounds case, produce a zero value.
+        let null_id = self.write_constant_null(result_type);
+
+        // Merge the results from the two paths.
+        let result_id = self.id_gen.next();
+        block.body.push(Instruction::phi(
+            result_type,
+            result_id,
+            &[(value_id, in_bounds_block), (null_id, header_block)],
+        ));
+
+        result_id
+    }
+
+    /// Emit code for bounds checks, per self.index_bounds_check_policy.
+    ///
+    /// Return a `BoundsCheckResult` indicating how the index should be
+    /// consumed. See that type's documentation for details.
+    #[allow(clippy::too_many_arguments)]
+    fn write_bounds_check(
+        &mut self,
+        ir_module: &crate::Module,
+        ir_function: &crate::Function,
+        fun_info: &FunctionInfo,
+        function: &mut Function,
+        base: Handle<crate::Expression>,
+        index: Handle<crate::Expression>,
+        block: &mut Block,
+    ) -> Result<BoundsCheckResult, Error> {
+        Ok(match self.index_bounds_check_policy {
+            IndexBoundsCheckPolicy::Restrict => {
+                match self.write_restricted_index(
+                    base,
+                    index,
+                    ir_module,
+                    ir_function,
+                    fun_info,
+                    function,
+                    block,
+                )? {
+                    KnownOrComputed::Known(restricted_index) => {
+                        BoundsCheckResult::KnownInBounds(restricted_index)
+                    }
+                    KnownOrComputed::Computed(restricted_index_id) => {
+                        BoundsCheckResult::Computed(restricted_index_id)
+                    }
+                }
+            }
+            IndexBoundsCheckPolicy::ReadZeroSkipWrite => {
+                match self.write_index_comparison(
```

Comment:
similarly, here? I think returning `BoundsCheckResult` would address the other concern I raised, if it's feasible